### ISSUE TITLE
Possible memroy leak when using ILU with BoomerAMG(?)

### ISSUE THREAD

**Issue by dpkn31** [hypre-space/hypre#961]
Dear hypre team, 

I am using  BoomerAMG via PETSc as a preconditioner for the PETSc GMRES solver. When using the ParILUK as a smoother, I am noticing a steady increase in memory usage and eventually, the program gets killed. This behaviour is not observed for any of the other smoothers.  Following the minimal example in the user-doc, the smoother is setup in PETSc: 

in  `src/ksp/pc/impls/hypre/hypre.c:`
```
HYPRE_BoomerAMGSetSmoothType(jac->hsolver, /*type  of complex smoother*/ 5); 
HYPRE_BoomerAMGSetSmoothNumLevels(jac->hsolver, /* number of levels*/ 25);
```
The version I am using is commit: 
`8f6bdc6ab` (tag: v2.29.0) Release 2.29.0 (#936)`

I am not using GPU acceleration. The issue is observed when running in serial(with single thread, and multiple threads), as well as with MPI (1 thread per process). 

Am I missing something here? The performance amd robustness of ParILUK way superior compared to PILUT and EUCLID, and I really would like to set this up as my 'default' setup! 

Thank you very much and best regards, 
dpk

**Comment by victorapm** [comment]
Hi @dpkn31, thank you for reaching out!

Can you reproduce this behavior by using any of the PETSc examples (drivers)?

**Comment by dpkn31** [comment]
Hi @victorapm , 

Sorry for the late reply, as advised I could NOT reproduce the leak from one of the existing examples from:
`src/ksp/ksp/tutorial/ex55.c` ( I am guessing the linear operator doesn't change during the iters, I haven't dug in to the details.).  Please note, with Petsc-3.19.5 the default hypre version is 2.28.0.

However, running my application with valgrind (with --leak-check=full) yields this: 

==8417== 4,194,304 bytes in 2 blocks are possibly lost in loss record 2,223 of 2,236 
==8417==    at 0x4A3F571: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8417==    by 0x124E1B7A: hypre_HostMalloc (memory.c:237)
==8417==    by 0x124E1CFD: hypre_MAlloc_core (memory.c:403)
==8417==    by 0x124E2259: hypre_CAlloc (memory.c:936)
==8417==    by 0x122907DF: hypre_SeqVectorInitialize_v2 (vector.c:97)
==8417==    by 0x1224C269: hypre_ParVectorInitialize_v2 (par_vector.c:128)
==8417==    by 0x1224C2AE: hypre_ParVectorInitialize (par_vector.c:142)
==8417==    by 0x123A4D59: hypre_ILUSetup (par_ilu_setup.c:396)
==8417==    by 0x120F4FE6: HYPRE_ILUSetup (HYPRE_parcsr_ilu.c:46)
==8417==    by 0x1239BB9E: hypre_BoomerAMGSetup (par_amg_setup.c:3657)
==8417==    by 0x120EFFD8: HYPRE_BoomerAMGSetup (HYPRE_parcsr_amg.c:53)
==8417==    by 0x1137C5AC: PCSetUp_HYPRE (hypre.c:420)
==8417== 
==8417== 4,194,304 bytes in 2 blocks are possibly lost in loss record 2,224 of 2,236 
==8417==    at 0x4A3A6A4: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8417==    by 0x124E1B8C: hypre_HostMalloc (memory.c:241)
==8417==    by 0x124E1CFD: hypre_MAlloc_core (memory.c:403)
==8417==    by 0x124E2222: hypre_MAlloc (memory.c:930)
==8417==    by 0x123A6D1F: hypre_ILUSetupMILU0 (par_ilu_setup.c:3628)
==8417==    by 0x123A698A: hypre_ILUSetupILU0 (par_ilu_setup.c:3501)
==8417==    by 0x123A9CE1: hypre_ILUSetupILUK (par_ilu_setup.c:4630)
==8417==    by 0x123A4F5D: hypre_ILUSetup (par_ilu_setup.c:451)
==8417==    by 0x120F4FE6: HYPRE_ILUSetup (HYPRE_parcsr_ilu.c:46)
==8417==    by 0x1239BB9E: hypre_BoomerAMGSetup (par_amg_setup.c:3657)
==8417==    by 0x120EFFD8: HYPRE_BoomerAMGSetup (HYPRE_parcsr_amg.c:53)
==8417==    by 0x1137C5AC: PCSetUp_HYPRE (hypre.c:420)
==8417== 
==8417== 4,195,804 (8 direct, 4,195,796 indirect) bytes in 1 blocks are definitely lost in loss record 2,225 of 2,236 
==8417==    at 0x4A3F571: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8417==    by 0x124E1B7A: hypre_HostMalloc (memory.c:237)
==8417==    by 0x124E1CFD: hypre_MAlloc_core (memory.c:403)
==8417==    by 0x124E2259: hypre_CAlloc (memory.c:936)
==8417==    by 0x12392C14: hypre_BoomerAMGSetup (par_amg_setup.c:955)
==8417==    by 0x120EFFD8: HYPRE_BoomerAMGSetup (HYPRE_parcsr_amg.c:53)
==8417==    by 0x1137C5AC: PCSetUp_HYPRE (hypre.c:420)
==8417==    by 0x113E89C7: PCSetUp (precon.c:994)
==8417==    by 0x110C3A9C: KSPSetUp (itfunc.c:406)
==8417==    by 0x110C799E: KSPSolve_Private (itfunc.c:824)
==8417==    by 0x110CAD9C: KSPSolve (itfunc.c:1070)
==8417==    by 0xE672223: operator()  ## from the 'solve' method of my app

To me it looks like memory for the ILU smoother is being continuously allocated, for each nonlinear solve(?)

Please note that the ILU  as smoother is not available in Petsc-3.19.5 yet,  [Here's](https://gitlab.com/dpk31/petsc/-/commit/effe12473f5a31d5116019ac9bb0f39195b21b3a) the diff from introducing it. 

Thanks for your time, 
deepak

**Comment by victorapm** [comment]
Thanks for the additional info!

~~Could you confirm which hypre commit did you use?~~

Edit: Sorry, now I see you used hypre v2.28.0. 

Would you be able to rerun with hypre's master? We've implemented a few bug fixes since v2.28.0 and this has been likely solved

**Comment by dpkn31** [comment]
Thank you very much! I can confirm that there are no memory leak issues with version: 57862ef6e (current master). 

Thanks and Best regards, 
Deepak
